### ISSUE TITLE
Changing image dynamically

### ISSUE THREAD

**Issue by appycamper** [xvrh/lottie-flutter#66]
Is it possible to swap images referenced in an animation dynamically? I'm trying to achieve something like described [here](https://stackoverflow.com/questions/49630767/how-do-i-change-the-a-single-image-in-the-lottie-json-file).

**Comment by xvrh** [comment]
I think this is already supported.
Most of the constructor have an `imageProviderFactory` parameter that allow to customise how to load the images.

```dart
Lottie.asset(
  'assets/example/data.json',
  imageProviderFactory: (image) {
    if (image.fileName == 'img_0') {
      return AssetImage('assets/image.png');
    } else if (image.fileName == 'img_1') {
      return NetworkImage('http://some_url/img.png');
    } else {
      // Fallback to the default image loading strategy
      return null;
    }
  },
),
```

**Comment by appycamper** [comment]
Thanks, I'm trying to change the images based on some user input but noticed that doing something like:

```
...
imageProviderFactory: (image) {
              if (image.fileName == 'img_0.png') {
                return AssetImage('assets/animations/example/images/$_image');
              } else {
                return null;
              }
```
Then later on:
```
setState(() {
   _image = 'img_0.png';
});
```
does not work, probably because the animation is cached. Is there a way to invalidate the cache or am I missing something? 

Many thanks

**Comment by xvrh** [comment]
Oh I see.
My answer doesn't work for this use case. The `imageProviderFactory` is, indeed, just for the initial loading.

What you can do is to change the `ui.Image` in the `LottieComposition.images[].loadedImage`:
```
composition.images['img_0'].loadedImage = yourNewImage;
```

Maybe I could add a "callback" so in the future we can write something like:

```dart
Lottie.asset(
  'assets/data.json',
  delegates: LottieDelegates(
    image: (composition, image) {
      if (image.id == 'img_0') {
        return myCustomImage;
      }
      
      // Will use the default method: composition.images[image.id].loadedImage;
      return null;
    },
  )
),
```

**Comment by appycamper** [comment]
Thanks works like a charm! Just wondering how you would go about saving or caching the composition so that I wouldn't have to redo the image replacements every time the animation is loaded?

**Comment by xvrh** [comment]
> how you would go about saving or caching the composition so that I wouldn't have to redo the image replacements every time the animation is loaded?

I'm not sure I understand this correctly. Do you have an example?

Once you have a `LottieComposition`, you can cache it and use it as many time as you want with the `Lottie(composition:)` constructor.

`LottieComposition` represents the loaded json and the `Lottie` widget is the player that displayed it on the screen.

**Comment by appycamper** [comment]
Thanks, so after making changes to the loaded images: 
```
composition.images['img_0'].loadedImage = yourNewImage;
```
would it be possible to serialise the composition and save it as a new json file?

**Comment by xvrh** [comment]
It is not possible to serialise a `LottieComposition`.

What you can do instead, is to use the standard dart json convert to read the file, modify it and write it back.
Something like that:

```dart
void _modifyImage() async {
  var data = await rootBundle.loadString('assets/data.json');
  var lottie = jsonDecode(data) as Map<String, dynamic>;
  
  for (var asset in lottie['assets']) {
    var assetMap = asset as Map<String, dynamic>;
    if (assetMap['id'] == 'image_0') {
      // You can also write the image encoded in base64 with "data:image/png;base64,..."
      assetMap['p'] = 'other_image.png';
    }
  }
  
  var modifiedLottie = jsonEncode(lottie);
  
  // Save it to a file at a place where you have write access
  File('xxx').writeAsStringSync(modifiedLottie);
}
```

**Comment by leonardocustodio** [comment]
> Oh I see.
> My answer doesn't work for this use case. The `imageProviderFactory` is, indeed, just for the initial loading.
> 
> What you can do is to change the `ui.Image` in the `LottieComposition.images[].loadedImage`:
> 
> ```
> composition.images['img_0'].loadedImage = yourNewImage;
> ```
> 
> Maybe I could add a "callback" so in the future we can write something like:
> 
> ```dart
> Lottie.asset(
>   'assets/data.json',
>   delegates: LottieDelegates(
>     image: (composition, image) {
>       if (image.id == 'img_0') {
>         return myCustomImage;
>       }
>       
>       // Will use the default method: composition.images[image.id].loadedImage;
>       return null;
>     },
>   )
> ),
> ```

LottieDelegates(image), doesn't exists anymore. Is there any alternative?

**Comment by xvrh** [comment]
@leonardocustodio can you share the `spinning_carrousel.zip` file so we can have a look?

If your images are embedded in the json file with `data:base64`, then [`imageProviderFactory` doesn't work](https://github.com/xvrh/lottie-flutter/blob/2360f643b7e8160bc81f9fa7f246693bc9ac30c0/lib/src/providers/lottie_provider.dart#L11). Maybe we can change that.

**Comment by leonardocustodio** [comment]
Here is the file:
[spinning_carrousel.zip](https://github.com/xvrh/lottie-flutter/files/5943116/spinning_carrousel.zip)

Thanks,

**Comment by xvrh** [comment]
@leonardocustodio 
I fixed a few things in this pull request: #130 
- Add the image callback in the `LottieDelegates` class
- Fix `imageProviderFactory` to work with a zip file.

This change has landed in the master branch which require a dart sdk >=2.12 (flutter beta). And I published `v0.8.0-nullsafety.3`

**Comment by sinshev** [comment]
@xvrh could you please add the image callback usage to examples?
